src/main/java/com/google/research/bleth/servlets/EnqueueSimulationServlet.java

```diff
@@ -0,0 +1,60 @@
+package com.google.research.bleth.servlets;
+
+import com.google.cloud.tasks.v2.AppEngineHttpRequest;
+import com.google.cloud.tasks.v2.CloudTasksClient;
+import com.google.cloud.tasks.v2.HttpMethod;
+import com.google.cloud.tasks.v2.QueueName;
+import com.google.cloud.tasks.v2.Task;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+ * A servlet used for enqueuing a task targeted at endpoint '/new-simulation',
+ * in order to create and run a new simulation.
+ */
+@WebServlet("/enqueue-simulation")
+public class EnqueueSimulationServlet extends HttpServlet {
+
+    static final String projectId = "bleth-2020";
+    static final String locationId = "europe-west1";
+    static final String queueId = "simulations-queue";
+
+    public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        try (CloudTasksClient client = CloudTasksClient.create()) {
+            // Construct the fully qualified queue name.
+            String queueName = QueueName.of(projectId, locationId, queueId).toString();
+
+            // Construct the HTTP request (for the task).
+            AppEngineHttpRequest httpRequest = AppEngineHttpRequest.newBuilder()
+                    .setRelativeUri("/new-simulation?" + toQueryString(request))
+                    .setHttpMethod(HttpMethod.POST)
+                    .build();
+
+            // Construct the task body.
+            Task task = Task.newBuilder()
+                    .setAppEngineHttpRequest(httpRequest)
+                    .build();
+
+            // Add the task to the queue.
+            client.createTask(queueName, task);
+        }
+
+        response.setContentType("text/html;");
+        response.getWriter().println("Task has been added to queue.");
+    }
+
+    private String toQueryString(HttpServletRequest request) {
```

Comment:
Why do you need this? Why not just use `request.getQueryString()`?
(AFAIK `request.getParameterMap()` just parses the query string.)

Reply:
This POST requests carries its parameters within the request body, which is why `request.getQueryString()` returns `null` in that case.

Reply:
In this case - why move the parameters from the body to the URL at all?

IMO it would be preferable to copy the POST body as-is.

```
AppEngineHttpRequest httpRequest = AppEngineHttpRequest.newBuilder()
                .setRelativeUri("/new-simulation?" + toQueryString(request))
                .setHttpMethod(HttpMethod.POST)
                .putHeaders("Content-Type", request.getContentType())
                .setPayload(getRequestPayload(request))
                .build();

// ...

  }

// ...

  private static byte[] getRequestPayload(HttpServletRequest request) throws IOException {
    try (InputStream s = request.getInputStream()) {
      return s.readAllBytes();
    }
  }
```

Reply:
I have three questions:
- If we pass the request parameters within its body, what do we need `putHeaders()` for? 
- Is there a difference between `setBody()` and `setPayload()`?
- If we do so, we don't have to add the query string at `setRelativeUri()`, right?

Reply:
* We need `putHeaders("Content-Type", ...)` since the body can only be interpreted in the context of a specific content-type.
* `setBody` is right, and `setPayload` is me being confused.
* Right - no need for a query string.